### PR TITLE
[CodeCompletion] Prioritize type matching overload for unresovled member

### PR DESCRIPTION
**Pull request by rintaro**
For exmaple:

```swift
    func foo(_: Int, _: IntOption)
    func foo(_: Float, _: FloatOption)

    foo(intVal, .<HERE>)
```
Previously, code completion suggests static member from `IntOption` and `FloatOption` without any prioritization. Prioritize members from `IntOption` because the user probably wants to input them.

In such cases, `CodeCompletionExpr` at the cursor position is pre-typehcecked to `IntOption`. So mark results with matching type with `ExprSpecific`.

rdar://problem/62121221

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test Linux

File: lib/IDE/CodeCompletion.cpp [hunk 1]
```diff
@@ -2080,6 +2078,23 @@ class CompletionLookup final : public swift::VisibleDeclConsumer {
...
+    if (!idealTy)
+      return false;
+    /// Consider optional object type is the ideal.
+    /// For exmaple:
+    ///   enum MyEnum { case foo, bar }
+    ///   func foo(_: MyEnum?)
+    ///   fooo(.<HERE>)
+    /// Prefer '.foo' and '.bar' over '.some' and '.none'.
+    idealTy = idealTy->lookThroughAllOptionalTypes();
+    return idealTy->isEqual(Ty);
```
Comment (benlangmuir) [inline lib/IDE/CodeCompletion.cpp:2095]: Why `isEqual` instead of `isConvertible`?

**Review by benlangmuir — APPROVED** [review]

File: lib/IDE/CodeCompletion.cpp (hunk 1 above)
Comment (rintaro) [reply lib/IDE/CodeCompletion.cpp:2095]: Returning convertible types by static member is pretty uncommon, I guess.
```swift
class Derived: Base {}
extension Base {
  static func createSub() -> Derived
}
```
I just didn't want to check them all by relatively heavy `isConvertible()` just for prioritization. But if you think it's worth it, I'm happy to do it.

Comment (benlangmuir) [reply lib/IDE/CodeCompletion.cpp:2095]: It's fine, just making sure I understand.

**Pull request merged**